cmd/acra-translator/server/server.go

```diff
@@ -193,129 +187,216 @@ func (server *ReaderServer) HandleConnectionString(parentContext context.Context
+				Errorln("Can't create HTTP decryptor")
+			errCh <- err
+			return
+		}
+		server.httpDecryptor = httpDecryptor
+		err = server.HandleHTTPConnection(httpContext, listener, server.config.IncomingConnectionHTTPString(), server.processHTTPConnection)
+		if err != nil {
+			// It is a "normal" case, when we shutdown service - we call 'cancel' on main level in SIGTERM/SIGINT handlers,
+			// so let's avoid treating context.Canceled as error here
+			if !errors.As(err, &context.Canceled) {
+				log.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorTranslatorCantHandleHTTPConnection).
+					Errorln("Took error on handling HTTP requests")
+				server.Stop()
+				errCh <- err
+			}
+			return
+		}
+	}()
+}
+
+func (server *ReaderServer) startGrpc(logger *log.Entry, decryptorData *common.TranslatorData, errCh chan<- error, listener net.Listener) {
+	server.backgroundWorkersSync.Add(1)
+	go func() {
+		defer server.backgroundWorkersSync.Done()
+		grpcLogger := logger.WithField(ConnectionTypeKey, GRPCConnectionType)
+		logger.WithField("connection_string", server.config.IncomingConnectionGRPCString()).Infof("Start process gRPC requests")
+		var err error
+		var opts []grpc.ServerOption
+		if server.config.WithTLS() {
+			opts = append(opts, grpc.Creds(credentials.NewTLS(server.config.GetTLSConfig())))
+		} else {
+			wrapper, err := network.NewSecureSessionConnectionWrapper(server.config.ServerID(), server.keystorage)
+			if err != nil {
+				grpcLogger.WithError(err).Errorln("Can't initialize Secure Session wrapper")
+				errCh <- err
+				return
+			}
+			opts = append(opts, grpc.Creds(wrapper))
+		}
+
+		server.listenerGRPC = listener
+		grpcListener := common.WrapListenerWithMetrics(listener)
+		grpcServer, err := server.grpcServerFactory.New(decryptorData, opts...)
+		if err != nil {
+			logger.WithError(err).Errorln("Can't create new grpc server")
+			errCh <- err
+			return
+		}
+		server.grpcServer = grpcServer
+		if err := grpcServer.Serve(grpcListener); err != nil {
+			grpcLogger.Errorf("failed to serve: %v", err)
+			server.Stop()
+			errCh <- err
+			return
+		}
+	}()
+}
+
+func (server *ReaderServer) detectPoisonRecords(poisonCallbackStorage *base.PoisonCallbackStorage) {
 	if server.config.DetectPoisonRecords() {
 		if server.config.ScriptOnPoison() != "" {
 			log.Infof("Add poison record callback with script execution %v", server.config.ScriptOnPoison())
-			poisonCallbacks.AddCallback(base.NewExecuteScriptCallback(server.config.ScriptOnPoison()))
+			poisonCallbackStorage.AddCallback(base.NewExecuteScriptCallback(server.config.ScriptOnPoison()))
 		}
 
 		// must be last
 		if server.config.StopOnPoison() {
 			log.Infoln("Add poison record callback with AcraTranslator termination")
-			poisonCallbacks.AddCallback(&base.StopCallback{})
+			poisonCallbackStorage.AddCallback(&base.StopCallback{})
 		}
 	}
+}
+
+func (server *ReaderServer) exitWithTimeout() {
```

Comment:
I'd call it something like `waitForExitTimeout` to avoid making an impression that calling this method *causes* the server to exit.

Reply:
ok, thank you